### ISSUE TITLE
sfdx force:source:push hanging on job ID after dependencies deploy

### ISSUE THREAD

**Issue by tcdahlberg** [SFDO-Tooling/CumulusCI#1579]
Windows Version 10.0.15063 Build 15063
Python 3.7.3
CumulusCI 3.7.0

Running 
`cci flow run dev_org --org dev  `

flow hangs on:
Running command: sfdx force:source:push -u test-rjaaao1g5seg@example.com
Job ID | 0Af0t00000SGJeGCAX

When I open the scratch org and look at the deploy status I can see the deploy succeeded. However no permissions were set for any of my objects or apps. Also none of my datasets are loaded.

I applied some of my managed permission sets to the User User and was able to get my application running (though not automated)

I thought this might be because of some branch work I was doing so I switched to my previously working master branch and that isn't working either.

Switched over to my laptop running CumulusCI 3.4.0 and was able to get a successful build.

My project:
https://github.com/tcdahlberg/Summit-Events-App

**Comment by davisagli** [comment]
@tcdahlberg Can you be more specific about what you mean by the "flow hangs"? Did it silently stop doing anything? Did it exit with an error, and if so what error?

Can you try running the command (`sfdx force:source:push -u test-rjaaao1g5seg@example.com`) on its own outside of CumulusCI and see if that gives any more clues? Maybe sfdx is outputting a message that CumulusCI is failing to relay to you.

**Comment by tcdahlberg** [comment]
No error. Just stopped at the Job ID and didn't progress. I have to end my terminal session to clear it out and be able to use cci.

the sfdx force:source:push works fine with no errors. There is nothing to push as the deployment when I look in the org says that it succeeded. The config_dev flow just never ran so objects and classes have no assigned permissions for System Admin. Tabs are not visible. The application I manage isn't visible. The dataset didn't load.

I just manually ran cci flow run config_dev and then everything worked. So, it is just not getting to that in the dev_org flow.

So, I am in a workable state at the moment but I don't have the flow automation I am used to.

Thanks for being so responsive!

**Comment by davisagli** [comment]
@tcdahlberg Oh right, running the push command a 2nd time is not the same scenario since there's nothing left to push. It sounds like maybe the first push is successfully submitting the package for deployment, but then getting stuck waiting for confirmation that it finished. Could you try the following variation in a fresh scratch org?
1. `cci flow run dependencies --org dev` to install EDA and get ready for pushing your package
2. `sfdx force:source:push -u SummitEventsDX__dev` (replace dev if you named the scratch org something else)
That way we can see if sfdx gives any additional info on the first push to the org.

It may also be worth double-checking that your installation of the Salesforce CLI is up to date.

**Comment by tcdahlberg** [comment]
@davisagli , The sfdx force:source:pushed worked fine with no errors.

My Salesforce CLI is up to date (7.47.0-0e5d18732f). On my desktop I tend to be fanatical about updates to my downfall sometimes. There is one warning when I update the CLI that I have to admit I have been ignoring because things just seem to work. 

`warning "salesforcedx > @salesforce/sfdx-plugin-lwc-test > @oclif/command@1.5.19" has unmet peer dependency "@oclif/plugin-help@^2".`

**Comment by davisagli** [comment]
@tcdahlberg hmm, ok, I'm not sure what's going on. I'll try to reproduce it with your project in Windows later this week.

**Comment by tcdahlberg** [comment]
@davisagli, thanks for the effort! I'm in a workable state at the moment with manually running through the flow so I can continue to develop.

**Comment by acrosman** [comment]
I'm running into this issue as well.  More or less the same setup as above although on cci 3.8.0, and sfdx 7.50.0-9030e49cbb win32-x64 node-v10.15.3 (been updating both since we first starting to see this hoping it would be resolved before I needed to pay attention).

cci runs the dev org flow until it does the `sfdx force:source:push` and then it returns a job Id and hangs.  If I cancel it (^-c), and re-run the same flow the process completes successfully.  I have at least one other colleague on Windows reporting the same problem with similar versions.

Oh, and @tcdahlberg that sfdx unmet peer dependency warning happens on nearly every extension; I'm guessing it's a flaw in a tutorial someplace. The developers don't see it because the peer is in their copy as a dev dependency. It's a simple fix if the maintainer is open to PRs for example: https://github.com/eltoroit/ETCopyData/pull/41

**Comment by tcdahlberg** [comment]
Thanks @acrosman, I did fix that unmet dependency, but am still getting the hang after the job id. I have taken to using sfdx:force:open to open the org and make sure that the job ID deployment has fully deployed. Than I have to terminate the cci terminal session, open another session, and run the _cci flow run config_dev --org dev_ to finish it up.

 I was going to try this all out on another machine to make sure my environment wasn't corrupted but haven't had an opportunity. I can confirm after updating CCI to 3.9.0 and I still have the issue.

**Comment by acrosman** [comment]
So far I've been able to confirm it's more complicated than just software versions or the project itself.

I had a colleague who was able to run our `cci flow run dev_org --org dev` successfully this morning.  He had newer Python (3.7.7) and newer Windows build (Version 1909 build 18363.720), but had CumulusCI 3.7.0. I've since updated both Windows and Python to match and continue to have the same problem.  

Is there a good way to get a listing of all the versions of the various dependencies of cci?

I am also wandering if there are reasonable ways to help debug the exchange of information between CCI and SFDX to help spot the error in more detail.

**Comment by tcdahlberg** [comment]
@acrosman , @davisagli 

I just set up my home computer for the obvious work from home reason. I used a newer version of python and windows then my work computer (which is tightly controlled when it comes to windows upgrades). I still experienced the hang on job ID after a fresh install of CumulusCI. My windows particulars:

Windows 10 version 1909 (10.0.18363 Build 18363)
Python 3.8.2
CumulusCI 3.9
sfdx-cli/7.50.0-9030e49cbb win32-x64 node-v10.15.3

**Comment by prescod** [comment]
@davisagli 

I'm not set up to debug Windows but if its useful I could jump on a call with someone and see if we can add tracing information together which will shed some light on this.

It's also possible that `cci flow run dev_org --org dev --debug` would offer some insight, but its a bit of a long-shot.

**Comment by acrosman** [comment]
I've been looking over package differences between my colleague who was able to get things to run smoothly and my own. I _may_ have found a solution, although I hit my daily scratch limit just after I got it to run once, so I'm not totally convinced yet (it'll take a little while to setup another org for additional testing).

We ran `pip list` and `pipx list` on both systems. There were several package differences so I started to run them one-by-one.  I'm not sure this _should_ make a difference, but figured it was worth a shot.

My copy of `userpath` was a version behind his, so I ran `pip install userpath --upgrade` and the next cycle of `cci flow run dev_org --org dev` worked.  I _think_ I had deleted the org before I run the install, but it's possible I hadn't (which is why I wanted to do a second test).

If `pip list` reports `userpath` as less than `1.3.0` you might try updating it.  I also updated `pipx` which added `click` before I updated `userpath` so it may have been the combination (or I may be just plain wrong here).

I did try `cci flow run dev_org --org dev --debug` but the extra information wasn't terribly enlightening.

My current packages:

```
> pip list
Package     Version
----------- --------
argcomplete 1.10.0
click       7.1.1
distro      1.4.0
entrypoints 0.3
flake8      3.7.8
mccabe      0.6.1
pip         20.0.2
pipx        0.15.1.3
pycodestyle 2.5.0
pyflakes    2.1.1
setuptools  41.2.0
userpath    1.3.0
```

**Comment by tcdahlberg** [comment]
@acrosman , my userpath is 1.3.0 and pipx is 0.15.1.3 so they match yours. I am still getting the hang on jobid.  

What's the format of your project? sfdx or mdapi? Mine is sfdx. Looking for clues.

I was curious when I ran debug a few lines before dx_push skipped update package xml and org_config.scratch. I don't pretend to understand the progression here:

```bash
2020-03-20 09:21:04: Skipping task update_package_xml (skipped when project_config.project__source_format != "sfdx" or not org_config.scratch)
2020-03-20 09:21:04: Skipping task deploy (skipped when project_config.project__source_format != "sfdx" or not org_config.scratch)
2020-03-20 09:21:04: ------------------------------------------------------------
2020-03-20 09:21:04: Running task: dx_push
2020-03-20 09:21:04: ------------------------------------------------------------
2020-03-20 09:21:04: 
2020-03-20 09:21:04: Running command:  sfdx force:source:push -u test-zlfsbhlv89bd@example.com
2020-03-20 09:21:04: Options:
2020-03-20 09:21:04:   command: sfdx force:source:push -u test-zlfsbhlv89bd@example.com
2020-03-20 09:21:04: Beginning task: SFDXOrgTask
2020-03-20 09:21:04: 
2020-03-20 09:21:04: Running command: sfdx force:source:push -u test-zlfsbhlv89bd@example.com
2020-03-20 09:21:10: Job ID | 0Af170000135w97CAA
```

@prescod  if you want to take on the task of troubleshooting I'd be willing to help with that. DM me on twitter if you are interested (same username on twitter).

**Comment by acrosman** [comment]
Ours is also an sfdx project and provides more or less the same output as yours. Since one of my colleagues is able to run the flow I believe it's related to dependencies some place not the project itself.  But I'm guessing here since I cannot concretely say I found a solution.

I did notice that my python instance is all kinds of complicated by having ended up with both a 64 and 32 bit version kicking around, and pipx was _really_ reluctant to switch over.  I had to uninstall cci, pipx, python37-32, delete pipx's settings in ~/.local, the 32-bit copy of all the libraries, and then it finally installed current versions of everything. From looking at the details it is pretty clear my previous setup was not clearly using one set of libraries I could trust the version numbers on.

Now I just need to get a new devhub setup so I can try to deploy again.

**Comment by acrosman** [comment]
Can confirm now that the update of `userpath` and all the other updates I've tried did not indeed resolve the problem for me other anyone else on our team who has a problem.

**Comment by prescod** [comment]
Sorry to hear that! I'll try to get a support plan together in a meeting I
have this afternoon and get back to you.

**Comment by prescod** [comment]
Can you guys think of any way in which unusual Unicode characters could be involved? I remember a ticket like this that boiled down to a failure to parse UTF-8 coming from a subprocess.

**Comment by davisagli** [comment]
@prescod hmm I can see how running in a terminal using cp1252 could lead to a UnicodeDecodeError when trying to decode the sfdx output as utf-8 (see cumulusci/tasks/command.py which the sfdx tasks are based on), but I don't see how it would simply hang.

**Comment by prescod** [comment]
Another hypothesis is that this relates simply to the amount of data being generated. If that's true, then a command like this might reproduce it:

cci task run command -o command 'python -c "print(__name__*100000000)"' > NUL:

(I think that's how to redirect to null on Windows)

**Comment by prescod** [comment]
@acrosman , @tcdahlberg :

Can you reproduce the bug with 

`cci task run dx_convert_from`

If so, that might be a faster reproduction step! Before you run this, make sure that your /src directory is non-existent, empty or easily reverted!

**Comment by acrosman** [comment]
`cci task run dx_convert_from` ran cleanly for me.  

`cci task run command -o command 'python -c "print(__name__*100000000)"' > NULL`  appears to do nothing, but does allow me to break it cleanly.

I'm not aware of any UTF-8 mb4 characters, or really any beyond the standard base Latin characters.

**Comment by davisagli** [comment]
@tcdahlberg @acrosman We think we've tracked this down, and we just released a possible fix in cumulusci 3.10.0. Could you try upgrading and see if it helps?

**Comment by acrosman** [comment]
I just installed 3.10.0 and it appears to be working successfully.
